source/source.go

```diff
@@ -0,0 +1,267 @@
+			Position:  sdk.Position(replayID),
+			CreatedAt: event.Data.Event.CreatedDate,
+			Metadata: map[string]string{
+				"channel":  event.Channel,
+				"replayId": replayID,
+				"action":   event.Data.Event.Type,
+			},
+		}, nil
+
+	case err := <-s.errors:
+		return sdk.Record{}, err
+
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	}
+}
+
+func (s *Source) Ack(_ context.Context, _ sdk.Position) error {
+	return nil // no ack needed
+}
+
+func (s *Source) Teardown(ctx context.Context) error {
+	// Unsubscribe
+	for _, pushTopicName := range s.config.PushTopicsNames {
+		unsubscribeResponse, err := s.streamingClient.UnsubscribeToPushTopic(ctx, pushTopicName)
+		if err != nil {
+			sdk.Logger(ctx).Warn().Msgf("unsubscribe error: failed to unsubscribe %q topic: %s", pushTopicName, err)
+		} else if !unsubscribeResponse.Successful {
+			sdk.Logger(ctx).Warn().Msgf("unsubscribe error: failed to unsubscribe %q topic: %s", pushTopicName, unsubscribeResponse.Error)
+		}
+	}
+
+	// Disconnect
+	disconnectResponse, err := s.streamingClient.Disconnect(ctx)
+	if err != nil {
+		return fmt.Errorf("disconnect error: %w", err)
+	}
+	if !disconnectResponse.Successful {
+		return fmt.Errorf("disconnect error: %s", disconnectResponse.Error)
+	}
+
+	s.subscriptions = nil
+	s.streamingClient = nil
+
+	return nil
+}
+
+// eventsWorker continuously queries for data updates from Salesforce
+func (s *Source) eventsWorker(ctx context.Context) {
+	defer close(s.events)
+
+	for {
+		// Receive event
+		connectResponse, err := s.streamingClient.Connect(ctx)
+		if err != nil {
+			s.errors <- fmt.Errorf("failed to receive event: %w", err)
+
+			return
+		}
+
+		// If not successful, check how to retry
+		if !connectResponse.Successful {
+			if nil == connectResponse.Advice {
+				s.errors <- fmt.Errorf("failed to receive event and no reconnection strategy provided by the server: %s", connectResponse.Error)
+
+				return
+			}
+
+			switch connectResponse.Advice.Reconnect {
+			case responses.AdviceReconnectRetry:
+				// Check if request can be retried
+				if connectResponse.Advice.Interval < 0 {
+					s.errors <- fmt.Errorf("server disallowed for reconnect, stopping")
+
+					return
+				}
+
+				// Wait and retry
+				time.Sleep(time.Millisecond * time.Duration(connectResponse.Advice.Interval))
+
+				continue
+
+			case responses.AdviceReconnectHandshake:
+				// Handshake and retry
+				if _, err := s.streamingClient.Handshake(ctx); err != nil {
+					s.errors <- fmt.Errorf("reconnect handshake error: %w", err)
+
+					return
+				}
+
+				continue
+
+			case responses.AdviceReconnectNone:
+				// Cannot retry
+				s.errors <- fmt.Errorf("server disallowed for reconnect, stopping")
+
+				return
+
+			default:
+				// Unexpected, cannot retry
+				s.errors <- fmt.Errorf("unsupported reconnect advice: %s", connectResponse.Advice.Reconnect)
+
+				return
+			}
+		}
+
+		// If successful, send event
+		for _, event := range connectResponse.Events {
```

Comment:
What happens when the connector starts for the first time? Will all the existing objects be pulled?

Reply:
No, only changes defined in Push Topic are reported.

Reply:
This one is actually related to the https://github.com/miquido/conduit-connector-salesforce/pull/1#discussion_r874689693. CometD returns Reply ID for incoming events. This theoretically can be used as a resume point when starting a connector. However, the order of events is not guaranteed. I need to check if this causes some events to be sent twice, which is not a desired behaviour for this connector I suppose.